### PR TITLE
Isaac/payment webhooks

### PR DESCRIPTION
**Pull request by IDubuque**

### CODE REVIEW

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 1]
```diff
@@ -269,6 +289,13 @@ function usePaymentsApi() {
       };
     }
 
+    if (options?.isSellerApiKey) {
+      return json as {
+        data: { data: any; decrypted_key: string };
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:294]: any way we can get the data type object? either hardcode or better if we can get it from the generated graphql

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 2]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+  sellerId: string;
+  url: string;
+  isProduction: boolean;
+  createdAt: Date;
+};
+
+export function usePaymentsWebhooksByAccountId(accountId: string) {
+  invariant(accountId, "accountId is required");
+
+  const address = useAddress();
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:904]: why do we need this? we don't use address for anything in the hook

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 3]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+          sellerId: webhook.seller_id,
+          url: webhook.url,
+          isProduction: webhook.is_production,
+          createdAt: new Date(webhook.created_at),
+        })) as PaymentsWebhooksType[];
+      } else {
+        return [] as PaymentsWebhooksType[];
+      }
+    },
+    { enabled: !!paymentsSellerId && !!address },
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:934]:
```suggestion
    { enabled: !!paymentsSellerId && !!accountId },
```

We should refetch on accountId change so we don't even have to get the address

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 4]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+        variables: {
+          sellerId: paymentsSellerId,
+        } as WebhooksBySellerIdQueryVariables,
+      });
+
+      if (error) {
+        console.error(error);
+      }
+
+      if (data && data.webhook.length > 0) {
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:922]: can be improved a bit if you do a ternary like I do in other places in the component, but that's a nit

`return data && data?.webhook.length > 0 ? data.webhook.map(... : [] as PaymentsWebhookType[]`

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 5]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+        return data.webhook.map((webhook) => ({
+          id: webhook.id,
+          sellerId: webhook.seller_id,
+          url: webhook.url,
+          isProduction: webhook.is_production,
+          createdAt: new Date(webhook.created_at),
+        })) as PaymentsWebhooksType[];
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:929]: the as for the data itself shouldn't be here, you need to give it in the part you did `data: any` before so it gets inherited

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 6]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+
+export type CreateWebhookInput = {
+  url: string;
+  isProduction: boolean;
+};
+
+export function usePaymentsCreateWebhook(accountId: string) {
+  invariant(accountId, "accountId is required");
+  const queryClient = useQueryClient();
+  const address = useAddress();
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:964]: again we don't need address here

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 7]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+  const { paymentsSellerId } = useApiAuthToken();
+
+  const [createWebhookBySellerId] = useInsertWebhookMutation({
+    refetchQueries: [WebhooksBySellerIdDocument],
+  });
+
+  return useMutationWithInvalidate(
+    async (input: CreateWebhookInput) => {
+      invariant(address, "No wallet address found");
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:973]: no need

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 8]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+      invariant(paymentsSellerId, "No seller id found");
+      invariant(isValidWebhookUrl(input.url), "Invalid webhook url");
+
+      const webhookInput: Webhook_Insert_Input = {
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:977]: nit: if you put the object directly inside the return you get the type safety anyway and you import one less thing, but this also works

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 9]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+export type UpdateWebhookInput = {
+  webhookId: string;
+  url?: string;
+  deletedAt?: Date;
+};
+
+export function usePaymentsUpdateWebhook(accountId: string) {
+  invariant(accountId, "accountId is required");
+  const queryClient = useQueryClient();
+  const address = useAddress();
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:1006]: same thing for the address

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 10]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+  const { paymentsSellerId } = useApiAuthToken();
+
+  const [updateWebhookBySellerId] = useUpdateWebhookMutation({
+    refetchQueries: [WebhooksBySellerIdDocument],
+  });
+
+  return useMutationWithInvalidate(
+    async (input: UpdateWebhookInput) => {
+      invariant(address, "No wallet address found");
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:1015]: same

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 11]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+      invariant(paymentsSellerId, "No seller id found");
+
+      return updateWebhookBySellerId({
+        variables: {
+          id: input.webhookId,
+          webhookValue: {
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:1021]: you did it here directly nice

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 12]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+            url: input.url,
+            deleted_at: input.deletedAt,
+          } as Webhook_Set_Input,
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:1024]: but in this case I don't think you need this? you should get the type safety already

File: @3rdweb-sdk/react/hooks/usePayments.ts [hunk 13]
```diff
@@ -862,3 +889,181 @@ export function usePaymentsSellerByAccountId(accountId: string) {
...
+        "api-key/get-decrypted-key",
+        {
+          sellerId: paymentsSellerId,
+        },
+        {
+          isSellerApiKey: true,
+        },
+      );
+    },
+    { enabled: !!paymentsSellerId && !!address },
```
Comment (nachoiacovino) [inline @3rdweb-sdk/react/hooks/usePayments.ts:1067]:
```suggestion
    { enabled: !!paymentsSellerId && !!accountId },
```

and remove useAddress

File: components/payments/settings/payment-webhooks.tsx [hunk 14]
```diff
@@ -7,5 +21,210 @@ interface PaymentsWebhooksProps {
...
+      variant: "solid",
+      title: header,
+      description,
+      status,
+      duration: 9000,
+      isClosable: true,
+    });
+  };
+
+  const { productionWebhooks, testnetWebhooks } = useMemo(() => {
```
Comment (nachoiacovino) [inline components/payments/settings/payment-webhooks.tsx:54]: this is a great improvement from before!

You can only return once if you have an optional param, this looks a bit cleaner to me but both work

`
  const { productionWebhooks, testnetWebhooks } = useMemo(
    () => ({
      productionWebhooks:
        webhooks?.filter((webhook) => webhook.isProduction) || [],
      testnetWebhooks:
        webhooks?.filter((webhook) => !webhook.isProduction) || [],
    }),
    [webhooks],
  );
`

File: components/payments/settings/payment-webhooks.tsx [hunk 15]
```diff
@@ -7,5 +21,210 @@ interface PaymentsWebhooksProps {
...
+      return { productionWebhooks, testnetWebhooks };
+    }
+    return { productionWebhooks: [], testnetWebhooks: [] };
+  }, [webhooks]);
+
+  const onUpdateWebhook: PaymentsWebhooksTableProps["onUpdate"] = (
+    webhook,
+    newUrl,
+  ) => {
+    if (!isValidWebhookUrl(newUrl)) {
```
Comment (nachoiacovino) [inline components/payments/settings/payment-webhooks.tsx:72]: can't we do this validation in the form itself? I'm doing something similar for successCallbackUrl inn create-update--checkout-button.tsx. You could copy the component as it or abtract it, (line 702-731)

It's always better if we can give feedback to the user _before_ they click a button to do something

File: components/payments/settings/payment-webhooks.tsx [hunk 16]
```diff
@@ -7,5 +21,210 @@ interface PaymentsWebhooksProps {
...
+        `${newUrl} is not a valid webhook url, please try a different url`,
+      );
+      return;
+    }
+
+    // send the request
+    updateWebhook(
+      { webhookId: webhook.id, url: newUrl },
+      {
+        onSuccess: () => {
```
Comment (nachoiacovino) [inline components/payments/settings/payment-webhooks.tsx:85]: on all onSuccess and onError, you should use the useTxNotifications hook we use everywhere else, with this you don't even have to define the toast yourself, so you only needs toasts for success and error, and the other ones you have for validation, you do in the form itself (check how we it in any other component we use react query)

Also, you need to import useTrack and track every user action, something like this in this case:
`              trackEvent({
                category: "payments",
                action: "update-webhook",
                label: "success",
              });`

File: components/payments/settings/payment-webhooks.tsx [hunk 17]
```diff
@@ -7,5 +21,210 @@ interface PaymentsWebhooksProps {
...
+          triggerAlert(
+            "success",
+            "Webhook Created",
+            `Successfully created  ${
+              webhook.isProduction ? "production" : "testnet"
+            } webhook: ${webhook.url}`,
+          );
+        },
+        onError: () => {
+          triggerAlert(
```
Comment (nachoiacovino) [inline components/payments/settings/payment-webhooks.tsx:95]: add the onError from useTxNotifications and tracking:

`              trackEvent({
                category: "payments",
                action: "delete-webhook",
                label: "error",
                error,
              });`

Comment (nachoiacovino) [reply components/payments/settings/payment-webhooks.tsx:95]: I also feel you could every easily merge delete and update if you wanted but that's something we can look over tomorrow, it's a nice code optimization

File: components/payments/settings/payment-webhooks.tsx [hunk 18]
```diff
@@ -7,5 +21,210 @@ interface PaymentsWebhooksProps {
...
+        },
+      },
+    );
+  };
+
+  const createWebhookHandlerFactory = (isProduction: boolean) => {
+    const onAddWebhook: PaymentsWebhooksTableProps["onCreate"] = async (
+      url,
+    ) => {
+      if (!isValidWebhookUrl(url)) {
```
Comment (nachoiacovino) [inline components/payments/settings/payment-webhooks.tsx:138]: same thing to do the validation in the form itself

File: components/payments/settings/payment-webhooks.tsx [hunk 19]
```diff
@@ -7,5 +21,210 @@ interface PaymentsWebhooksProps {
...
+            ) : (
+              <CodeBlock code={webhookApiKey?.data?.decrypted_key ?? ""} />
+            )
+          }
+        />
+
+        <Divider />
+
+        <Flex flexDir="column" gap={2}>
+          <Heading size={"title.md"}>Production</Heading>
```
Comment (nachoiacovino) [inline components/payments/settings/payment-webhooks.tsx:204]: nit, always nice if you only have a string to not use brackets
```suggestion
          <Heading size="title.md">Production</Heading>
```

File: components/payments/settings/payments-webhooks-table.tsx [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+);
+
+interface EditTableRowProps {
+  webhook: PaymentsWebhooksType;
+  onEdit: (webhook: PaymentsWebhooksType, newUrl: string) => void;
+  onDelete: (webhook: PaymentsWebhooksType) => void;
+}
+
+const EditTableRow = ({ webhook, onEdit, onDelete }: EditTableRowProps) => {
+  const [editUrl, setEditUrl] = useState(webhook.url ?? "");
```
Comment (nachoiacovino) [inline components/payments/settings/payments-webhooks-table.tsx:88]: I still wonder if we really need state here for this but we can look over it tomorrow, we might do but we might find a cleaner way

File: components/payments/settings/payments-webhooks-table.tsx [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+    </Tr>
+  );
+};
+
+interface CreateTableRowProps {
+  onCreate: (webhookUrl: string) => void;
+}
+
+const CreateTableRow = ({ onCreate }: CreateTableRowProps) => {
+  const [createUrl, setCreateUrl] = useState("");
```
Comment (nachoiacovino) [inline components/payments/settings/payments-webhooks-table.tsx:170]: same for this

File: components/payments/settings/payments-webhooks-table.tsx [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+    onCreate(createUrl);
+    setCreateUrl("");
+  };
+
+  return (
+    <Tr borderBottomWidth={1} _last={{ borderBottomWidth: 0 }}>
+      <Td borderBottomWidth="inherit" borderBottomColor="accent.100">
+        <UrlInputField
+          webhookUrl={createUrl}
+          onChange={setCreateUrl}
```
Comment (nachoiacovino) [inline components/payments/settings/payments-webhooks-table.tsx:183]: see here would makes more sense to use useForm even if it's a simple thing (and then we don't need useState) and we let useForm handle all the logic and validation

File: components/payments/settings/payments-webhooks-table.tsx [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+
+export const PaymentsWebhooksTable: React.FC<PaymentsWebhooksTableProps> = ({
+  webhooks,
+  onUpdate,
+  onDelete,
+  onCreate,
+  isLoading,
+  isFetched,
+}) => {
+  const [webhookToRevoke, setWebhookToRevoke] =
```
Comment (nachoiacovino) [inline components/payments/settings/payments-webhooks-table.tsx:204]: same thing here happy to look at this tomorrow more in depth

**Review by nachoiacovino — CHANGES_REQUESTED** [review]
More important thing: fixing the build

All in all, amazing improvement compared to yesterday, if you start seeing me ask for nit stuff, it means I'm starting to not have enough things to complain about :D

let's pair on this tomorrow to bring it over the finish line!

File: components/payments/settings/payments-webhooks-table.tsx (hunk 20 above)
Comment (IDubuque) [reply components/payments/settings/payments-webhooks-table.tsx:88]: I changed this from a ref because renders were clearing all the input fields

**Review by nachoiacovino — APPROVED** [review]
let's merge so we can test and let's improve it tomorrow with the things we went over in the call!

**Pull request merged**